usenix89/Utilities/Soundex/brachman/sp.1
.TH SP 1-LOCAL "12 December 1986"
.UC 4
.SH NAME
sp \- give possible spellings
.br
mksp \- maintain sp dictionaries
.br
calcsoundex \- calculate soundex values
.SH SYNOPSIS
.B sp
[
.B -vace
] [
.B -f dictionary-list
] [
.B word ...
]
.br
.B mksp
[
.B -adt
] [
.B -v#
]
.B dictionary
.br
.B calcsoundex
[
-v
]
.SH DESCRIPTION
.I Sp
takes one or more words as input and for each word prints
a list of possible spellings.
If the words are not given on the command line, the program prompts and reads
from the standard input.
You must know the first letter of the word.
Upper case is mapped to lower case.
Words must start with an alphabetic character, but any subsequent
characters need not be alphabetic (but see the limitation below on words that
may appear in the dictionary).
Blanks are allowed within a word.
.PP
Up to ten dictionaries previously created by
.I mksp
may be specified by a command line argument and an environment variable.
The name of a dictionary is specified by a pathname,
not including the suffix (.dir or .pag).
A list of dictionaries consists of one or more colon separated dictionary
names.
The environment variable SPPATH may be set to a dictionary list.
If a command line dictionary list is given in addition to the SPPATH variable,
all dictionaries are used.
If no dictionaries are specified the program looks for default dictionaries.
.PP
To reduce the size of the word list, certain heuristics are used.
Normally, all words
.I sp
considers to be a "satisfactory" match are printed.
The \fB-c\fR option causes only close
matches or an exact match to be printed.
The \fB-e\fR option only prints an exact match.
The \fB-a\fR option causes all words matched to be printed.
The output is sorted alphabetically and indicators are printed beside each
word:
.sp 2
.in +0.5i
.nf
.na
   X == exact match
.br
   ! == close match
.br
   * == good match
.br
 ' ' == matched
.in -0.5i
.fi
.ad
.sp 2
Duplicated words are not removed from the listing.
.PP
If the \fB-v\fR flag is given,
.I sp
becomes verbose.
.PP
.I Mksp
is a program to maintain dictionaries for use with
.I sp.
The \fB-a\fR option is used to create a new dictionary or to add
words to an existing dictionary.
The words to be put in the dictionary are read from the standard input,
one per line.
The \fB-v\fR flag (which may immediately be followed by an optional number)
causes some information to be printed as words are processed.
A non-flag argument to
.I mksp
is assumed to be the prefix of the name of the dictionary files.
The dictionary consists of two files, one with a ".dir" suffix and one with
a ".pag" suffix (see \fBdbm\fR(3X)).
If these files do not exist,
.I mksp
will create both.
The words need not be sorted.
There should not be duplicates in the word list when creating a new dictionary
but when words are added to an existing dictionary, duplicates are ignored.
Upper case letters are stored in the dictionary but
.I mksp
maps upper case to lower case when checking for duplicate words.
.I Sp
is case insensitive when searching the database.
The first character of a word must be alphabetic and
the second character (if present) must be either alphabetic, a single quote,
an ampersand, a period, or a blank.
There is no restriction on the third and subsequent characters.
.PP
The \fB-d\fR option is used to delete words from the specified dictionary.
The words are read from the standard input.
If a word is not found in the dictionary, no message is printed.
The comparison is case insensitive.
.PP
The \fB-t\fR option prints the contents of the specified dictionary.
The words are not sorted.
.PP
.I Calcsoundex
reads words from the standard input (one per line) and prints the soundex
code corresponding to each word on stdout.
With the \fB-v\fR flag,
.I calcsoundex
also echoes each word.
.SH EXAMPLE
.in +0.5i
.na
.nf
% mksp -a mydictionary
aardvark
precipitation
<more words>
zyzz
<^D>
% sp -c -f mydictionary:herdictionary
Word? propogate
!  1. perfect
!  2. perfectible
<more words>
!  7. propagate
!  8. proposition
<more words>
Word? <^D>
.fi
.ad
.in -0.5i
.SH FILES
.nf
.na
<dictionary.pag>, <dictionary.dir>        dictionary data base
/usr/public/lib/sp/{soundex,aux.soundex}  default dictionaries
.fi
.ad
.SH LIMITATIONS
No more than 10 dictionaries may be specified.
The maximum length of a word is 50 characters.
The program can return up to 400 matches taking up a maximum of 20480 bytes.
The limitation on what the second character of a word can be is due to
the algorithm used to compress the dictionaries; there is some room in the
data structure for a few other characters to become valid.
Dbm doesn't work between a Sun and VAX across NFS so you can't share a
dictionary between a VAX and a Sun.
.PP
There is a limit on the number of words having the same soundex code that can
appear in a single dictionary.
This value should be at least 256 (on a VAX/Sun it is 1024), but it depends
on how the program has been configured locally.
If you come up against this limitation you can split your dictionary; e.g.,
extract every second word from the big dictionary to make a new dictionary,
then delete the words from the big dictionary.
The following pipeline can be used to determine the number of times each
soundex code appears in a list of words (one per line):
.sp 2
.ti +0.5i
calcsoundex | sort | uniq -c | sort -r -0n
.SH "SEE ALSO"
spell(1), uniq(1), dbm(3X), ndbm(3)
.br
Donald E. Knuth, The Art of Computer Programming, Vol. 3,
Sorting and Searching, Addison-Wesley, pp. 391-392, 1973.
.SH AUTHOR
Barry Brachman
.br
Dept. of Computer Science
.br
University of British Columbia
.SH BUGS
You may not agree on what constitutes a match.
You are likely to have to create your own dictionary as the UNIX
dictionary is far from complete.  In particular, the suffixes
have been removed from most words.
The limitations mentioned above are arbitrary.
The limitation on the second character of a word is disgusting.



usenix89/Utilities/Soundex/brachman/sp.c
/* vi: set tabstop=4 : */

/*
 * Version 1.3 December 1986
 *
 * sp - spell word
 *
 * Usage:	sp [-f dictionary-list] [-eavc] [word ...]
 *
 * Compute the Soundex code for each word on the command line
 * (or each word on the standard input) and compare against a
 * dictionary
 *
 * The soundex dictionary list may be specified on the command line
 * The environment variable SPPATH may be set to a list of colon
 * separated pathnames of soundex dictionaries.
 * If a command line dictionary-list (a colon separated list of pathnames) is
 * given in addition to the SPPATH variable, all dictionaries are used.
 *
 * To reduce the size of the word list, certain heuristics are used:
 * the -a option causes all words matched to be printed
 * The output is alphabetically sorted and indicators are printed
 * beside each word:
 *	X   == exact match
 *	!   == close match
 *	*   == near match
 * ' '  == matched
 *
 * Note that the maximum number of colliding words is MAXCOUNT due to the
 * data structure used.
 *
 * Permission is given to copy or distribute this program provided you
 * do not remove this header or make money off of the program.
 *
 * Please send comments and suggestions to:
 *
 * Barry Brachman
 * Dept. of Computer Science
 * Univ. of British Columbia
 * Vancouver, B.C. V6T 1W5
 *
 * .. {ihnp4!alberta, uw-beaver}!ubc-vision!ubc-cs!brachman
 * brachman@cs.ubc.cdn
 * brachman%ubc.csnet@csnet-relay.arpa
 * brachman@ubc.csnet
 */

#include <sys/types.h>
#include <sys/file.h>
#include <ctype.h>
#include <stdio.h>

#ifdef NEWDBM
#include <ndbm.h>
#else !NEWDBM
#include <dbm.h>
#endif NEWDBM

#include "sp.h"

#define streq(X, Y)	(!strcmp(X, Y))
#define range(S)	((strlen(S) + 4) / 5)

#define USAGE		"Usage: sp [-f dictionary-list] [-eavc] [word ...]"

char word[MAXWORDLEN + 2];

datum FETCH();

char *fileptr[MAXDICT + 1];	/* Up to MAXDICT dictionaries + sentinel */
int dict_ptr = 0;

char *wordptr[MAXWORDS], *wordlistptr;
char wordlist[WORDSPACE];
int nmatched;

/*
 * Soundex codes
 * The program depends upon the numbers zero through six being used
 * but this can easily be changed
 */
char soundex_code_map[26] = {
/***	 A  B  C  D  E  F  G  H  I  J  K  L  M  N  O  P	***/ 
		 0, 1, 2, 3, 0, 1, 2, 0, 0, 2, 2, 4, 5, 5, 0, 1,

/***	 Q  R  S  T  U  V  W  X  Y  Z			***/
		 2, 6, 2, 3, 0, 1, 0, 2, 0, 2
};

int aflag, cflag, eflag, vflag;

main(argc, argv)
int argc;
char **argv;
{
	register int fflag, i;
	register char *p;
	char *getenv();

	argc--; argv++;
	fileptr[0] = (char *) NULL;
	while (argc > 0 && argv[0][0] == '-') {
		fflag = 0;		/* to break out of following loop... */
		for (i = 1; argv[0][i] != '\0' && fflag == 0; i++) {
			switch (argv[0][i]) {
			case 'a':
				aflag = 1;
				break;
			case 'c':
				cflag = 1;
				break;
			case 'e':
				eflag = 1;
				break;
			case 'f':
				if (argc == 1) {
					fprintf(stderr, "%s\n", USAGE);
					exit(1);
				}
				mkfilelist(argv[1]);
				argc--;
				argv++;
				fflag = 1;		/* break out of loop */
				break;
			case 'v':
				vflag = 1;
				break;
			default:
				fprintf(stderr, "%s\n", USAGE);
				exit(1);
			}
		}
		argc--, argv++;
	}

	if ((p = getenv("SPPATH")) != (char *) NULL)
		mkfilelist(p);
	if (fileptr[0] == (char *) NULL)
		mkfilelist(DEFAULT_SPPATH);
	if (vflag) {
		printf("Using dictionaries:\n");
		for (i = 0; fileptr[i] != (char *) NULL; i++)
			if (strlen(fileptr[i]) > 0)
				printf("\t%s\n", fileptr[i]);
	}
	if (argc) {
		for (i = 0; i < argc; i++) {
			if (!eflag)
				printf("%s:\n", argv[i]);
			apply(argv[i]);
			if (!eflag)
				printf("\n");
		}
	}
	else {
		int ch, len;

		while (1) {
			printf("Word? ");
			if (fgets(word, sizeof(word), stdin) == (char *) NULL) {
				printf("\n");
				break;
			}
			len = strlen(word);
			if (word[len - 1] != '\n') {
				fprintf(stderr, "sp: Word too long: %s", word);
				while ((ch = getchar()) != '\n')	/* flush rest of line */
					putc(ch, stderr);
				putc('\n', stderr);
				continue;
			}
			word[--len] = '\0';
			if (len > MAXWORDLEN) {
				fprintf(stderr, "sp: Word too long: %s\n", word);
				continue;
			}

			apply(word);
			if (!eflag)
				printf("\n");
		}
	}
}

/*
 * Apply the Soundex search for a word to each dictionary in turn
 * Note that 'DBMINIT' opens both the '.dir' and the '.pag' files
 * and we must close them to avoid running out of file descriptors
 *
 * This routine gets called each time a word is looked up and therefore
 * the dbm files may be repeatedly opened and closed.  Since the vast majority
 * of the time this program is invoked for just a single word it doesn't seem
 * worthwhile to do the right thing by saving file descriptors/DBM pointers.
 * There probably won't be more than two dictionaries in use anyway.
 */
apply(word)
char *word;
{
	register int code, i, nodicts;

	nmatched = 0;
	wordlistptr = wordlist;
	if ((code = soundex(word, 3)) == BAD_WORD)
		return;
	nodicts = 1;
	for (i = 0; fileptr[i] != (char *) NULL; i++) {
		if (strlen(fileptr[i]) == 0)
			continue;
		if (DBMINIT(fileptr[i], O_RDONLY) != -1) {
			proc(code);
			nodicts = 0;
		}
		DBMCLOSE();
	}
	if (nodicts) {
		fprintf(stderr, "sp: Can't open any dictionaries\n");
		exit(1);
	}
	if (vflag && !eflag && nmatched == 0)
		printf("%s: no match\n", word);
	else
		choose(word);
}

/*
 * Look the word up in the current dictionary
 * and save all the matches
 * Note that only three digits are of the Soundex code are stored
 * in a dictionary
 */
proc(soundex)
int soundex;
{
	register int c, len;
	datum dbm_key, dbm_content;
	key_t *key, keyvec[KEYSIZE];
	char *mk_word(), *p;

	key = keyvec;
	dbm_key.dptr = (char *) key;
	dbm_key.dsize = KEYSIZE;
	c = 0;
	while (1) {
		mk_key(key, soundex, c);
		dbm_content = FETCH(dbm_key);

		if (dbm_content.dptr == 0)
			break;

		if (IS_DELETED(dbm_content)) {
			if (++c > MAXCOUNT) {
				fprintf(stderr, "sp: entry count overflow\n");
				exit(1);
			}
			continue;
		}

		if (nmatched == MAXWORDS) {
			fprintf(stderr, "sp: Too many matches\n");
			exit(1);
		}

		p = mk_word(dbm_content.dptr, dbm_content.dsize, soundex);
		len = strlen(p);
		if (wordlistptr + len >= &wordlist[WORDSPACE]) {
			fprintf(stderr, "sp: Out of space for words\n");
			exit(1);
		}
		strncpy(wordlistptr, p, len);
		wordlistptr[len] = '\0';
		wordptr[nmatched++] = wordlistptr;
		wordlistptr += len + 1;
		if (++c > MAXCOUNT) {
			fprintf(stderr, "sp: entry count overflow\n");
			exit(1);
		}
	}
}

/*
 * Select and print those words which we consider
 * to have matched 'word'
 */
choose(word)
register char *word;
{
	register int c, code, i, len, mcount, wordlen;
	register char *p;
	int compar();

	code = soundex(word, 4);
	qsort(wordptr, nmatched, sizeof(char *), compar);
	c = range(word);
	wordlen = strlen(word);
	mcount = 0;
	for (i = 0; i < nmatched; i++) {
		p = wordptr[i];
		if (strmatch(word, p) == 0) {
			printf("X");
			if (eflag) {
				printf(" %s\n", word);
				return;
			}
		}
		else if (eflag)
			continue;
		else if (soundex(p, 4) == code)
			printf("!");
		else if (aflag &&
			(wordlen < (len = strlen(p)) - c || len > wordlen + c))
			printf(" ");
		else if (!cflag)
			printf("*");
		else
			continue;
		printf("%3d. %s\n", mcount + 1, p);
		mcount++;
	}
	if (vflag)
		printf("(%d total matches)\n", nmatched);
}

/*
 * Compute an 'n' digit Soundex code for 'word' 
 * See mksp.c
 */
soundex(word, n)
register char *word;
int n;
{
	register int c, digit_part, previous_code, soundex_length;
	register char *p, *w;
	char wcopy[MAXWORDLEN + 2];

	strcpy(wcopy, word);
	p = w = wcopy;
	while (*p != '\0') {
		if (isupper(*p))
			*p = tolower(*p);
		p++;
	}
	if (!isalpha(*w)) {
		fprintf(stderr, "sp: Improper word: %s\n", word);
		return(BAD_WORD);
	}
	digit_part = 0;
	soundex_length = 0;
	previous_code = soundex_code_map[*w - 'a'];
	for (p = w + 1; *p != '\0' && soundex_length < n; p++) {
		if (!isalpha(*p))
			continue;
		c = soundex_code_map[*p - 'a'];
		if (c == 0 || previous_code == c) {
			previous_code = c;
			continue;
		}
		digit_part = digit_part * 7 + c;
		previous_code = c;
		soundex_length++;
	}
	while (soundex_length++ < n)
		digit_part *= 7;
	return((digit_part << 5) + *w - 'a');
}

/*
 * Process a path string (environment variable SPPATH, DEFAULT_SPPATH, or an
 * arg) by separating the pathnames into strings pointed to by elements
 * of 'fileptr'
 * End of list indicated by fileptr entry of NULL
 *
 * No attempt made to ignore duplicate pathnames
 */
mkfilelist(p)
register char *p;
{
	register int len;
	register char *path, *start;
	char *malloc();

	while (*p != '\0' && dict_ptr < MAXDICT) {
		start = p;
		while (*p != ':' && *p != '\0')
			p++;
		if (start == p && *p == ':') {	/* colon with nothing else */
			p++;
			continue;
		}
		len = p - start;
		path = (char *) malloc((unsigned) (len + 1));
		if (path == (char *) NULL) {
			fprintf(stderr, "sp: Out of dictionary space\n");
			exit(1);
		}
		strncpy(path, start, len);
		path[len] = '\0';
		fileptr[dict_ptr++] = path;
	}
	fileptr[dict_ptr] = (char *) NULL;
}

compar(p, q)
char **p, **q;
{

	return(strmatch(*p, *q));
/*	return(strcmp(*p, *q)); */	/* use if you prefer case sensitive */
}

